<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="generator" content="litedown 0.7">
<title>New Resources for Working with NEON Data</title>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}
body, .abstract, code, .footnotes, footer, #refs, .caption { font-size: .9em; }
li li { font-size: .95em; }
ul:has(li > input[type="checkbox"]) { list-style: none; padding-left: 1em; }
*, :before, :after { box-sizing: border-box; }
a { color: steelblue; }
pre, img { max-width: 100%; }
pre { white-space: pre-wrap; word-break: break-word; }
pre code { display: block; padding: 1em; overflow-x: auto; }
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre, th) > code, code[class], div > .caption { background: #f8f8f8; }
pre > code:is(:not([class]), .language-plain, .language-none, .plain), .box, .figure, .table { background: inherit; border: 1px solid #eee; }
pre > code {
  &.message { border-color: #9eeaf9; }
  &.warning { background: #fff3cd; border-color: #fff3cd; }
  &.error { background: #f8d7da; border-color: #f8d7da; }
}
.fenced-chunk { border-left: 1px solid #666; }
.code-fence {
  opacity: .4;
  border: 1px dashed #666;
  border-left: 2px solid;
  &:hover { opacity: inherit; }
}
.box, .figure, .table, table { margin: 1em auto; }
div > .caption { padding: 1px 1em; }
.figure { p:has(img, svg), pre:has(svg) { text-align: center; } }
.flex-col { display: flex; justify-content: space-between; }
table {
  &:only-child:not(.table > *) { margin: auto; }
  th, td { padding: 5px; font-variant-numeric: tabular-nums; }
  thead, tfoot, tr:nth-child(even) { background: whitesmoke; }
  thead th { border-bottom: 1px solid #ddd; }
  &:not(.datatable-table) {
    border-top: 1px solid #666;
    border-bottom: 1px solid #666;
  }
}
blockquote {
  color: #666;
  margin: 0;
  padding: 1px 1em;
  border-left: .5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC {
  a { text-decoration: none; }
  ul { list-style: none; padding-left: 1em; }
  & > ul { padding: 0; }
  ul ul { border-left: 1px solid lightsteelblue; }
}
.body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.main-number::after { content: "."; }
span[class^="ref-number-"] { font-weight: bold; }
.ref-number-fig::after, .ref-number-tab::after { content: ":"; }
.cross-ref-chp::before { content: "Chapter "; }
.cross-ref-sec::before { content: "Section "; }
.cross-ref-fig::before, .ref-number-fig::before { content: "Figure "; }
.cross-ref-tab::before, .ref-number-tab::before { content: "Table "; }
.cross-ref-eqn::before, .MathJax_ref:has(mjx-mtext > mjx-c + mjx-c)::before { content: "Equation "; }
.abstract, #refs {
  &::before { display: block; margin: 1em auto; font-weight: bold; }
}
.abstract::before { content: "Abstract"; text-align: center; }
#refs::before { content: "Bibliography"; font-size: 1.5em; }
.ref-paren-open::before { content: "("; }
.ref-paren-close::after { content: ")"; }
.ref-semicolon::after { content: "; "; }
.ref-and::after { content: " and "; }
.ref-et-al::after { content: " et al."; font-style: italic; }
.footnote-ref a {
  &::before { content: "["; }
  &::after { content: "]"; }
}
section.footnotes {
  margin-top: 2em;
  &::before { content: ""; display: block; max-width: 20em; }
}
.fade {
  background: repeating-linear-gradient(135deg, white, white 30px, #ddd 32px, #ddd 32px);
  opacity: 0.6;
}

@media print {
  body { max-width: 100%; }
  tr, img { break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  body:not(.pagesjs) pre:has(.line-numbers):not(:hover) { white-space: pre; }
}
</style>
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>New Resources for Working with NEON Data</h1></div>
</div>
<div class="body">
<p>This tutorial provides a tour of recent (as of fall 2025) updates and enhancements to NEON code resources.</p>
<p>If you are new to using NEON data, we recommend first following the <a href="https://www.neonscience.org/resources/learning-hub/tutorials/download-explore-neon-data" target="_blank">Download and Explore</a> tutorial to get familiar with NEON data formats and the <code>neonUtilities</code> package.</p>
<h2 id="sec:set-up-r-environment">Set up R environment</h2>
<p>First install and load the necessary packages.</p>
<pre><code># install packages. can skip this step if 

# the packages are already installed

install.packages(&quot;neonUtilities&quot;)

install.packages(&quot;lubridate&quot;)

install.packages(&quot;devtools&quot;)

devtools::install_github(&quot;NEONScience/NEON-geolocation/geoNEON&quot;)



# load packages

library(neonUtilities)

library(geoNEON)

library(ggplot2)

library(lubridate)
</code></pre>
<h2 id="sec:the-neonos-package">The neonOS package</h2>
<p>The <code>neonOS</code> package provides functions specifically for working with NEON observational (OS) data products. NEON’s human-collected data present some unique challenges: the structure and data relationships within each product are unique, automation is more difficult, and sample relationships are often essential. The <code>neonOS</code> package contains functions for duplicate checking, table joining, and sample exploration.</p>
<p>For a detailed introduction to the <code>neonOS</code> package, see the <a href="https://www.neonscience.org/resources/learning-hub/tutorials/neonos-duplicates-joins" target="_blank">full tutorial</a>. Here we’ll do a quick overview.</p>
<p>We’ll use soil data throughout as an example dataset to illustrate code features. Let’s say we’re interested in the relationships between soil chemistry, microbial communities, and disturbance events. Start by downloading Soil physical and chemical properties, periodic (DP1.10086.001) from Soaproot Saddle (SOAP), 2018-2024.</p>
<h3 id="sec:feature-omit-progress-bars">Feature: Omit progress bars</h3>
<p>Progress bars can now be suppressed in download functions in the <code>neonUtilities</code> package. This can be helpful when running code in an automated workflow or batch script. Note that messages associated with progress will be suppressed along with the bars, but messages about exceptions or errors will still be printed to the console.</p>
<pre><code>soil &lt;- loadByProduct(dpID=&quot;DP1.10086.001&quot;, 
                     site=&quot;SOAP&quot;,
                     startdate=&quot;2018-01&quot;,
                     enddate=&quot;2024-12&quot;,
                     include.provisional=T,
                     progress=F,
                     check.size=F)



list2env(soil, .GlobalEnv)

## &lt;environment: R_GlobalEnv&gt;
</code></pre>
<p>To start, we want to assess whether soil moisture is associated with different nitrogen content. Moisture data are in the <code>sls_soilMoisture</code> table, and nitrogen data are in the <code>sls_soilChemistry</code> table. We’ll need to join the tables.</p>
<p>But before joining, let’s check both tables for duplicates.</p>
<h3 id="sec:feature-check-for-duplicate-records">Feature: Check for duplicate records</h3>
<pre><code>moisturedups &lt;- removeDups(sls_soilMoisture, 
                           variables_10086)

## No duplicated key values found!

chemdups &lt;- removeDups(sls_soilChemistry,
                       variables_10086)

## No duplicated key values found!
</code></pre>
<p>Neither table contains any duplicates. This is common - staff internal to NEON take advantage of these code resources, too, and use them to detect and resolve duplicates. However, especially if you use Provisional data, it’s a good idea to check. See the <code>neonOS</code> <a href="https://www.neonscience.org/resources/learning-hub/tutorials/neonos-duplicates-joins" target="_blank">tutorial</a> for more details about the <code>removeDups()</code> function, and to see an example that does contain duplicates.</p>
<p>Now we can join the tables.</p>
<h3 id="sec:feature-join-observational-data-tables">Feature: Join observational data tables</h3>
<pre><code>moistchem &lt;- joinTableNEON(sls_soilMoisture,
                           sls_soilChemistry)
</code></pre>
<p>Now that we have the nitrogen and moisture data in a single table, we can compare.</p>
<pre><code>plot(nitrogenPercent~soilMoisture, 
     data=moistchem,
     pch=20)
</code></pre>
<p><img src="https://raw.githubusercontent.com/NEONScience/NEON-Data-Skills/main/tutorials/R/NEON-general/neon-overview/new-features-2025/rfigs/plot-chem-moist-1.png" alt=" " /></p>
<p>Within this dataset, wetter soils tend to also be higher in nitrogen.</p>
<h2 id="sec:event-data">Event data</h2>
<p>We are also interested in the impact of disturbance on these soils. Soaproot Saddle is in California - has it burned during the observed time period?</p>
<p>The Site management and event reporting (DP1.10111.001) data product contains records of disturbance and management events that occur at NEON sites. We could download all records from SOAP, but a new function in <code>neonUtilities</code> allows us to run a targeted query for specific event types.</p>
<h3 id="sec:feature-get-event-data-by-event-type">Feature: Get event data by event type</h3>
<pre><code>sim &lt;- byEventSIM(&quot;fire&quot;, site=&quot;SOAP&quot;)



sim$sim_eventData

##                                    uid domainID siteID namedLocation
## 1 2c8c0bf8-caf0-4c48-bad3-590d0e737047      D17   SOAP          SOAP
## 2 0c1fcf9f-5bee-478c-8412-0bf57925a04b      D17   SOAP          SOAP
## 3 4ffe7ac8-ff68-48e8-97e6-2efd446cd3da      D17   SOAP          SOAP
## 4 53b0b814-8441-41c4-a44b-91fc3413d97b      D17   SOAP          SOAP
## 5 6b641a93-0494-4998-9d8a-c56df5d70891      D17   SOAP          SOAP
## 6 8bc10c21-79b3-4421-b0ae-703f98067e6f      D17   SOAP          SOAP
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 locationID
## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             SOAP_007.basePlot.all, SOAP_016.basePlot.all, SOAP_017.basePlot.all, SOAP_018.basePlot.all, SOAP_019.basePlot.all, SOAP_003.tickPlot.tck, SOAP_036.mosquitoPoint.mos, SOAP_039.mosquitoPoint.mos, SOAP_041.mosquitoPoint.mos
## 2 SOAP_002.basePlot.all, SOAP_005.basePlot.all, SOAP_006.basePlot.all, SOAP_008.basePlot.all, SOAP_012.basePlot.all, SOAP_013.basePlot.all, SOAP_023.basePlot.all, SOAP_024.basePlot.all, SOAP_026.basePlot.all, SOAP_027.basePlot.all, SOAP_028.basePlot.all, SOAP_043.basePlot.all, SOAP_044.basePlot.all, SOAP_046.basePlot.all, SOAP_050.basePlot.all, SOAP_051.basePlot.all, SOAP_052.basePlot.all, SOAP_055.basePlot.all, SOAP_058.basePlot.all, SOAP_059.basePlot.all, SOAP_060.basePlot.all, SOAP_061.basePlot.all, SOAP_005.tickPlot.tck, SOAP_006.tickPlot.tck, SOAP_026.tickPlot.tck, SOAP_002.mammalGrid.mam, SOAP_006.mammalGrid.mam, SOAP_008.mammalGrid.mam, SOAP_012.mammalGrid.mam, SOAP_032.mammalGrid.mam, SOAP_033.mosquitoPoint.mos, SOAP_035.mosquitoPoint.mos, SOAP_037.mosquitoPoint.mos, SOAP_038.mosquitoPoint.mos, SOAP_039.mosquitoPoint.mos, SOAP_040.mosquitoPoint.mos, SOAP
## 3 SOAP_002.basePlot.all, SOAP_005.basePlot.all, SOAP_006.basePlot.all, SOAP_008.basePlot.all, SOAP_012.basePlot.all, SOAP_013.basePlot.all, SOAP_023.basePlot.all, SOAP_024.basePlot.all, SOAP_026.basePlot.all, SOAP_027.basePlot.all, SOAP_028.basePlot.all, SOAP_043.basePlot.all, SOAP_044.basePlot.all, SOAP_046.basePlot.all, SOAP_050.basePlot.all, SOAP_051.basePlot.all, SOAP_052.basePlot.all, SOAP_055.basePlot.all, SOAP_058.basePlot.all, SOAP_059.basePlot.all, SOAP_060.basePlot.all, SOAP_061.basePlot.all, SOAP_005.tickPlot.tck, SOAP_006.tickPlot.tck, SOAP_026.tickPlot.tck, SOAP_002.mammalGrid.mam, SOAP_006.mammalGrid.mam, SOAP_008.mammalGrid.mam, SOAP_012.mammalGrid.mam, SOAP_032.mammalGrid.mam, SOAP_033.mosquitoPoint.mos, SOAP_035.mosquitoPoint.mos, SOAP_037.mosquitoPoint.mos, SOAP_038.mosquitoPoint.mos, SOAP_039.mosquitoPoint.mos, SOAP_040.mosquitoPoint.mos, SOAP
## 4 SOAP_002.basePlot.all, SOAP_005.basePlot.all, SOAP_006.basePlot.all, SOAP_008.basePlot.all, SOAP_012.basePlot.all, SOAP_013.basePlot.all, SOAP_023.basePlot.all, SOAP_024.basePlot.all, SOAP_026.basePlot.all, SOAP_027.basePlot.all, SOAP_028.basePlot.all, SOAP_043.basePlot.all, SOAP_044.basePlot.all, SOAP_046.basePlot.all, SOAP_050.basePlot.all, SOAP_051.basePlot.all, SOAP_052.basePlot.all, SOAP_055.basePlot.all, SOAP_058.basePlot.all, SOAP_059.basePlot.all, SOAP_060.basePlot.all, SOAP_061.basePlot.all, SOAP_005.tickPlot.tck, SOAP_006.tickPlot.tck, SOAP_026.tickPlot.tck, SOAP_002.mammalGrid.mam, SOAP_006.mammalGrid.mam, SOAP_008.mammalGrid.mam, SOAP_012.mammalGrid.mam, SOAP_032.mammalGrid.mam, SOAP_033.mosquitoPoint.mos, SOAP_035.mosquitoPoint.mos, SOAP_037.mosquitoPoint.mos, SOAP_038.mosquitoPoint.mos, SOAP_039.mosquitoPoint.mos, SOAP_040.mosquitoPoint.mos, SOAP
## 5 SOAP_002.basePlot.all, SOAP_005.basePlot.all, SOAP_006.basePlot.all, SOAP_008.basePlot.all, SOAP_012.basePlot.all, SOAP_013.basePlot.all, SOAP_023.basePlot.all, SOAP_024.basePlot.all, SOAP_026.basePlot.all, SOAP_027.basePlot.all, SOAP_028.basePlot.all, SOAP_043.basePlot.all, SOAP_044.basePlot.all, SOAP_046.basePlot.all, SOAP_050.basePlot.all, SOAP_051.basePlot.all, SOAP_052.basePlot.all, SOAP_055.basePlot.all, SOAP_058.basePlot.all, SOAP_059.basePlot.all, SOAP_060.basePlot.all, SOAP_061.basePlot.all, SOAP_005.tickPlot.tck, SOAP_006.tickPlot.tck, SOAP_026.tickPlot.tck, SOAP_002.mammalGrid.mam, SOAP_006.mammalGrid.mam, SOAP_008.mammalGrid.mam, SOAP_012.mammalGrid.mam, SOAP_032.mammalGrid.mam, SOAP_033.mosquitoPoint.mos, SOAP_035.mosquitoPoint.mos, SOAP_037.mosquitoPoint.mos, SOAP_038.mosquitoPoint.mos, SOAP_039.mosquitoPoint.mos, SOAP_040.mosquitoPoint.mos, SOAP
## 6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     SOAP_014.basePlot.all, SOAP_016.basePlot.all, SOAP_020.basePlot.all, SOAP_045.basePlot.all, SOAP_047.basePlot.all, SOAP_053.basePlot.all, SOAP_054.basePlot.all, SOAP_056.basePlot.all, SOAP_062.phenology.phe, SOAP
##    startDate    endDate ongoingEvent estimatedOrActualDate dateRemarks            eventID
## 1 2020-02-24 2020-03-09            N                  &lt;NA&gt;        &lt;NA&gt; SOAP.20200224.fire
## 2 2020-09-04 2020-10-04            Y                actual        &lt;NA&gt; SOAP.20200904.fire
## 3 2020-10-04 2020-11-03            Y                actual        &lt;NA&gt; SOAP.20201004.fire
## 4 2020-11-03 2020-12-03            Y                actual        &lt;NA&gt; SOAP.20201004.fire
## 5 2020-12-03 2020-12-24            N                actual        &lt;NA&gt; SOAP.20201004.fire
## 6 2021-06-29 2021-07-08            N                actual        &lt;NA&gt; SOAP.20210629.fire
##   samplingProtocolVersion eventType    methodTypeChoice name scientificName otherScientificName
## 1       NEON.DOC.003282vC      fire fire-controlledBurn &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
## 2       NEON.DOC.003282vD      fire       fire-wildfire &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
## 3       NEON.DOC.003282vD      fire       fire-wildfire &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
## 4       NEON.DOC.003282vD      fire       fire-wildfire &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
## 5       NEON.DOC.003282vD      fire       fire-wildfire &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
## 6       NEON.DOC.003282vD      fire       fire-wildfire &lt;NA&gt;           &lt;NA&gt;                &lt;NA&gt;
##   fireSeverity biomassRemoval minQuantity maxQuantity quantityUnit reporterType         remarks
## 1      unknown           &lt;NA&gt;          NA          NA         &lt;NA&gt;    secondary controlled burn
## 2         high           &lt;NA&gt;          NA          NA         &lt;NA&gt;      primary      Creek Fire
## 3         high           &lt;NA&gt;          NA          NA         &lt;NA&gt;      primary      Creek Fire
## 4         high           &lt;NA&gt;          NA          NA         &lt;NA&gt;      primary      Creek Fire
## 5         high           &lt;NA&gt;          NA          NA         &lt;NA&gt;      primary      Creek Fire
## 6         high           &lt;NA&gt;          NA          NA         &lt;NA&gt;      primary       Blue Fire
##            recordedBy dataQF
## 1 0000-0002-4748-8985   &lt;NA&gt;
## 2 0000-0001-7920-7757   &lt;NA&gt;
## 3 0000-0001-7920-7757   &lt;NA&gt;
## 4 0000-0001-7920-7757   &lt;NA&gt;
## 5 0000-0001-7920-7757   &lt;NA&gt;
## 6 0000-0001-7920-7757   &lt;NA&gt;
</code></pre>
<p>There are six records about fires at SOAP. Looking at the details, we can see one controlled burn in 2020, the Creek Fire in 2020 (four records), and the Blue Fire in 2021.</p>
<p>How does this compare to the measurement dates for the chemistry data?</p>
<pre><code>unique(year(sls_soilChemistry$collectDate))

## [1] 2018 2024
</code></pre>
<p>Chemistry measurements were made in 2018 and 2024, before and after the fires. Do nitrogen values differ between the two measurement dates?</p>
<pre><code>plot(nitrogenPercent~soilMoisture, 
     data=moistchem[which(year(moistchem$collectDate.y)==2018),],
     pch=20)

points(nitrogenPercent~soilMoisture, 
     data=moistchem[which(year(moistchem$collectDate.y)==2024),],
     pch=20, col=&quot;red&quot;)
</code></pre>
<p><img src="https://raw.githubusercontent.com/NEONScience/NEON-Data-Skills/main/tutorials/R/NEON-general/neon-overview/new-features-2025/rfigs/plot-time-1.png" alt=" " /></p>
<p>No obvious differences in the pattern between the two measurement dates.</p>
<p>If we wanted to dig into the effects of fire more deeply, note that the event data table includes a <code>locationID</code> field. This field contains a record of the NEON sampling plots affected by the event - in this case, which plots burned. If there are both burned and unburned plots in our dataset, this could be very informative.</p>
<h3 id="sec:pending-feature-site-management-location-data-details">Pending feature: Site management location data details</h3>
<p>There is a NEON code package in development that will help users to more easily work with the location data associated with site events. Stay tuned!</p>
<h2 id="sec:community-data">Community data</h2>
<p>We mentioned above that we are also interested in microbial communities. What microbes are living in these soils, and how have they changed over time? Let’s download the Soil microbe community taxonomy (DP1.10081.002) data for the same time period at SOAP.</p>
<h3 id="sec:feature-seamless-stacking-of-microbe-community-data">Feature: Seamless stacking of microbe community data</h3>
<p>The community data are in the expanded data package.</p>
<pre><code>micro &lt;- loadByProduct(dpID=&quot;DP1.10081.002&quot;, 
                     site=&quot;SOAP&quot;,
                     startdate=&quot;2018-01&quot;,
                     enddate=&quot;2024-12&quot;,
                     package=&quot;expanded&quot;,
                     include.provisional=T,
                     progress=F,
                     check.size=F)

## 
## NEON.D17.SOAP.DP1.10081.002.2018-05.expanded.20250129T000730Z.RELEASE-2025.zip could not be downloaded. Re-attempting.

## Stacking per-sample files. These files may be very large; download data in smaller subsets if performance problems are encountered.

## Variables file was not found or was inconsistent for table mct_soilPerSampleTaxonomy_ITS. Schema will be inferred; performance may be reduced.

list2env(micro, .GlobalEnv)

## &lt;environment: R_GlobalEnv&gt;
</code></pre>
<p>Let’s look at the distribution by phylum of fungal taxon units at SOAP.</p>
<pre><code>gg &lt;- ggplot(mct_soilPerSampleTaxonomy_ITS, aes(phylum, individualCount)) +
  geom_col() +
  theme(axis.text.x = element_text(angle = 90))

gg
</code></pre>
<p><img src="https://raw.githubusercontent.com/NEONScience/NEON-Data-Skills/main/tutorials/R/NEON-general/neon-overview/new-features-2025/rfigs/plot-microbe-data-1.png" alt=" " /></p>
<p>But can we compare the communities at the two time points?</p>
<pre><code>unique(year(mct_soilSampleMetadata_ITS$collectDate))

## [1] 2018
</code></pre>
<p>Marker gene data were collected in 2024, but the community taxonomy data are not available yet.</p>
<p>Suppose we wanted to compare soil chemistry data to plant communities instead of microbial communities? The Plant presence and percent cover (DP1.10058.001) data product would be a good place to start.</p>
<h3 id="sec:pending-feature-neonplants-r-package">Pending feature: neonPlants R package</h3>
<p>We are very close to releasing the <code>neonPlants</code> package, which will facilitate use of NEON’s various plant data products, including both community and biomass data. Stay tuned!</p>
<h2 id="sec:samples">Samples</h2>
<h2 id="sec:other-languages-and-platforms">Other languages and platforms</h2>
<h3 id="sec:feature-python-version-of-neonutilities">Feature: Python version of neonutilities</h3>
<p>For years, neonUtilities has only been available in R, but as of 2024 we have a Python version!</p>
<p>Aside from some minor formatting differences - <code>neonutilities</code> in all lower case, function names in <code>snake_case</code> rather than <code>camelCase</code> - the functions and inputs are the same in Python as in R. For example, to download the same soil dataset:</p>
<pre><code>
import neonutilities as nu

soil = nu.load_by_product(dpid=&quot;DP1.10086.001&quot;, 
                     site=&quot;SOAP&quot;,
                     startdate=&quot;2018-01&quot;,
                     enddate=&quot;2024-12&quot;,
                     include_provisional=True,
                     progress=False,
                     check_size=False)

</code></pre>
<p>Other code packages, including <code>neonOS</code> and <code>geoNEON</code>, haven’t been implemented in Python yet. Similarly, not all features of neonUtilities are available in Python yet. In particular, we are actively working on incoporating the surface-atmosphere exchange functions in Python.</p>
<h3 id="sec:pending-feature-surface-atmosphere-exchange-data-stacking-in-python">Pending feature: Surface-atmosphere exchange data stacking in Python</h3>
<p>This should be available within a few weeks.</p>
<h3 id="sec:feature-cloud-mode">Feature: Cloud mode</h3>
<p>The <code>loadByProduct()</code> function works by downloading NEON data files to a temporary directory and then reading them into R in stacked form. With new cloud-native software, it is possible to accomplish this without the intermediate download step - the files are imported directly to the code environment and stacked simultaneously. If you are working in a cloud environment, add the input <code>cloud.mode=T</code> (or <code>cloud_mode=True</code> in Python) to <code>loadByProduct()</code> for much faster and more efficient data access.</p>
<p>If you are new to working in cloud environments but want to start exploring, we recommend checking out <a href="https://cyverse.org/" target="_blank">CyVerse</a>.</p>
<h2 id="sec:stay-connected">Stay connected!</h2>
<p>New code releases and updates are announced via NEON <a href="https://www.neonscience.org/data-samples/data-notifications" target="_blank">data notifications</a>. Sign up to get data notifications sent to your inbox by making a NEON <a href="https://www.neonscience.org/about/user-accounts" target="_blank">user account</a>.</p>
<p>If you have any questions, or want to request new features, use the <a href="https://www.neonscience.org/about/contact-us" target="_blank">Contact Us</a> page.</p>
</div>
</body>
</html>
